hazelcast/src/main/java/com/hazelcast/instance/DefaultAddressPicker.java

```diff
@@ -269,28 +277,55 @@ private static AddressDefinition pickLoopbackAddress(String host, int defaultPor
     AddressDefinition pickMatchingAddress(Collection<InterfaceDefinition> interfaces) throws SocketException {
         Enumeration<NetworkInterface> networkInterfaces = NetworkInterface.getNetworkInterfaces();
         boolean preferIPv4Stack = preferIPv4Stack();
-        boolean matchInterfaceDefinition = CollectionUtil.isNotEmpty(interfaces);
+        boolean preferIPv6Addresses = preferIPv6Addresses();
+        AddressDefinition matchingAddress = null;
+
+        // There are 3 possible value pairs for preferIPv4Stack & preferIPv6Addresses:
+        // - preferIPv4Stack=true, preferIPv6Addresses=false: Only an IPv4 address will be picked.
+        // - preferIPv4Stack=false, preferIPv6Addresses=false: Either an IPv4 or IPv6 address may be picked, no preference.
+        // - preferIPv4Stack=false, preferIPv6Addresses=true: Either an IPv4 or IPv6 address may be picked
+        // but IPv6 address will be preferred over IPv4.
+
         while (networkInterfaces.hasMoreElements()) {
             NetworkInterface ni = networkInterfaces.nextElement();
-            if (!matchInterfaceDefinition && skipInterface(ni)) {
+            if (isEmpty(interfaces) && skipInterface(ni)) {
                 continue;
             }
             Enumeration<InetAddress> e = ni.getInetAddresses();
             while (e.hasMoreElements()) {
                 InetAddress inetAddress = e.nextElement();
                 if (preferIPv4Stack && inetAddress instanceof Inet6Address) {
+                    // IPv4 stack is preferred, so only IPv4 address can be picked.
                     continue;
                 }
-                if (matchInterfaceDefinition) {
-                    AddressDefinition address = match(inetAddress, interfaces);
-                    if (address != null) {
-                        return address;
+
+                AddressDefinition address = getMatchingAddress(interfaces, inetAddress);
+                if (address == null) {
+                    continue;
+                }
+                matchingAddress = address;
+
+                if (preferIPv6Addresses) {
+                    // IPv6 address is preferred, return if address is IPv6.
+                    if (inetAddress instanceof Inet6Address) {
+                        return matchingAddress;
                     }
-                } else if (!inetAddress.isLoopbackAddress()) {
-                    return new AddressDefinition(inetAddress);
+                } else if (inetAddress instanceof Inet4Address) {
+                    // No IPv6 address preference, return if address is IPv4.
+                    return matchingAddress;
                 }
             }
         }
+        // nothing matched to IP version preference, return what we have.
+        return matchingAddress;
```

Comment:
Minor: why not return the 1st match instead of the last in the case of no-match to version preference?
It seems more natural to me, when i would list my box interfaces/addresses to see the one at the top selected.